src/DynamoCoreWpf/ViewModels/PackageManager/InstalledPackagesViewModel.cs

```diff
@@ -1,51 +1,261 @@
+        /// </summary>
+        internal void PopulateFilters()
         {
-            dynamoViewModel.PackageManagerClientViewModel.GoToWebsite();
+            var currentSelection = CurrentFilterSelection;
+            Filters.Clear();
+            var filterNames = Model.LocalPackages
+                .Where(pkg =>
+                    !(pkg.LoadState.ScheduledState == PackageLoadState.ScheduledTypes.None &&
+                      pkg.LoadState.State == PackageLoadState.StateTypes.None))
+                .Select(pkg => new PackageViewModel(dynamoViewModel, pkg)).Select(vm => vm.PackageLoadStateText)
+                .Distinct();
+            Filters.AddRange(filterNames.Select(f => new PackageFilter(f, this)));
+
+            if (Filters.Any())
+            {
+                Filters.Insert(0, new PackageFilter(this));
+            }
+
+            ResetCurrentSelection(currentSelection);
+
+            RaisePropertyChanged(nameof(Filters));
         }
 
-        private void InitializeLocalPackages()
+        private string CurrentFilterSelection => Filters.FirstOrDefault(f => f.IsChecked)?.Name;
+
+        private bool NoActiveFilterSelected
         {
-            foreach (var pkg in Model.LocalPackages)
+            get
             {
-                LocalPackages.Add(new PackageViewModel(this.dynamoViewModel, pkg));
+                var currentFilter = CurrentFilterSelection;
+                return currentFilter == null || currentFilter == Resources.PackageFilter_Name_All;
             }
+        }
+
+        private bool IsPackageVisible(Package pkg)
+        {
+            var currentFilter = CurrentFilterSelection;
+            var pkgViewModel = new PackageViewModel(dynamoViewModel, pkg);
+            return NoActiveFilterSelected || pkgViewModel.PackageLoadStateText == currentFilter;
+        }
 
-            this.Model.PackageAdded += (pkg) =>
+        private PackageViewModel NewPackageViewModel(Package pkg)
+        {
+            var viewModel = new PackageViewModel(dynamoViewModel, pkg);
+            viewModel.Model.PropertyChanged += PackageModelOnPropertyChanged;
+            return viewModel;
+        }
+
+        private void ClearPackageViewModel(PackageViewModel viewModel)
+        {
+            viewModel.Model.PropertyChanged -= PackageModelOnPropertyChanged;
+        }
+
+        private void ClearPackagesViewModel()
+        {
+            foreach (var viewModel in LocalPackages)
             {
-                LocalPackages.Add(new PackageViewModel(dynamoViewModel, pkg));
-                RaisePropertyChanged("LocalPackages");
+                ClearPackageViewModel(viewModel);
+            }
+
+            LocalPackages.Clear();
+        }
+
+        private void InitializeLocalPackages()
+        {
+            ClearPackagesViewModel();
+            LocalPackages.AddRange(Model.LocalPackages.Select(NewPackageViewModel));
+
+            Model.PackageAdded += (pkg) =>
```

Comment:
I think we should try to clean these handlers up if possible... packageLoader may live longer than the `InstalledPackagesViewModel` - maybe - I can't remember if it lives for the application lifetime or just when the prefs view is opened... Also need to consider test contexts.

Reply:
It is for the application lifetime as it stands right now.